### ISSUE TITLE
Watch of individual objects is effectively broken

### ISSUE THREAD

**Issue by invino4** [GoogleCloudPlatform/kubernetes#6480]
Using a watch expression on endpoints can lead to unpredictable behavior that includes returning a watch stream that ignores filters (e.g. a {name} determination) and returns items of the wrong type (e.g. changes to PODS in the ENDPOINTS stream).  

# Repro Steps

```
curl http://localhost:8080/api/v1beta1/watch/endpoints/guestbook?namespace=default
```

returns 

```
{"type":"ADDED","object":{"kind":"Endpoints","id":"kubernetes","uid":"b0f8c701-d7df-11e4-88b2-a0481c7e9013","creationTimestamp":"2015-03-31T19:54:07Z","selfLink":"/api/v1beta1/endpoints/kubernetes?namespace=default","resourceVersion":7,"apiVersion":"v1beta1","namespace":"default","protocol":"TCP","endpoints":["172.31.9.40:6443"],"subsets":[{"addresses":[{"IP":"172.31.9.40"}],"ports":[{"port":6443,"protocol":"TCP"}]}]}}
{"type":"ADDED","object":{"kind":"Endpoints","id":"kubernetes-ro","uid":"b0fe3434-d7df-11e4-88b2-a0481c7e9013","creationTimestamp":"2015-03-31T19:54:07Z","selfLink":"/api/v1beta1/endpoints/kubernetes-ro?namespace=default","resourceVersion":8,"apiVersion":"v1beta1","namespace":"default","protocol":"TCP","endpoints":["172.31.9.40:7080"],"subsets":[{"addresses":[{"IP":"172.31.9.40"}],"ports":[{"port":7080,"protocol":"TCP"}]}]}}
{"type":"ADDED","object":{"kind":"Endpoints","id":"redis-master","uid":"e75b5a4e-d7f2-11e4-88b2-a0481c7e9013","creationTimestamp":"2015-03-31T22:11:39Z","selfLink":"/api/v1beta1/endpoints/redis-master?namespace=default","resourceVersion":4066,"apiVersion":"v1beta1","namespace":"default","protocol":"TCP","endpoints":["172.17.0.17:6379"],"targetRefs":[{"endpoint":"172.17.0.17:6379","targetRef":{"kind":"Pod","namespace":"default","name":"redis-master-controller-pflbd","uid":"9d0ee66a-d7f1-11e4-88b2-a0481c7e9013","resourceVersion":"3815"}}],"subsets":[{"addresses":[{"IP":"172.17.0.17","targetRef":{"kind":"Pod","namespace":"default","name":"redis-master-controller-pflbd","uid":"9d0ee66a-d7f1-11e4-88b2-a0481c7e9013","resourceVersion":"3815"}}],"ports":[{"port":6379,"protocol":"TCP"}]}]}}
{"type":"ADDED","object":{"kind":"Endpoints","id":"redis-slave","uid":"52c784ab-d7f3-11e4-88b2-a0481c7e9013","creationTimestamp":"2015-03-31T22:14:39Z","selfLink":"/api/v1beta1/endpoints/redis-slave?namespace=default","resourceVersion":4178,"apiVersion":"v1beta1","namespace":"default","protocol":"TCP","endpoints":["172.17.0.18:6379","172.17.0.19:6379"],"targetRefs":[{"endpoint":"172.17.0.18:6379","targetRef":{"kind":"Pod","namespace":"default","name":"redis-slave-controller-iebcg","uid":"0b95b8b0-d7f3-11e4-88b2-a0481c7e9013","resourceVersion":"4114"}},{"endpoint":"172.17.0.19:6379","targetRef":{"kind":"Pod","namespace":"default","name":"redis-slave-controller-aiuwh","uid":"0b95c218-d7f3-11e4-88b2-a0481c7e9013","resourceVersion":"4122"}}],"subsets":[{"addresses":[{"IP":"172.17.0.18","targetRef":{"kind":"Pod","namespace":"default","name":"redis-slave-controller-iebcg","uid":"0b95b8b0-d7f3-11e4-88b2-a0481c7e9013","resourceVersion":"4114"}},{"IP":"172.17.0.19","targetRef":{"kind":"Pod","namespace":"default","name":"redis-slave-controller-aiuwh","uid":"0b95c218-d7f3-11e4-88b2-a0481c7e9013","resourceVersion":"4122"}}],"ports":[{"port":6379,"protocol":"TCP"}]}]}}
{"type":"MODIFIED","object":{"kind":"Endpoints","id":"guestbook","uid":"d036af74-d7f3-11e4-88b2-a0481c7e9013","creationTimestamp":"2015-03-31T22:18:09Z","selfLink":"/api/v1beta1/endpoints/guestbook?namespace=default","resourceVersion":253567,"apiVersion":"v1beta1","namespace":"default","protocol":"TCP","endpoints":["172.17.0.26:3000","172.17.0.30:3000","172.17.0.32:3000"],"targetRefs":[{"endpoint":"172.17.0.26:3000","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-k5zf8","uid":"89cc6848-d7fd-11e4-88b2-a0481c7e9013","resourceVersion":"6422"}},{"endpoint":"172.17.0.30:3000","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-7muj6","uid":"26e07ca5-da6d-11e4-88b2-a0481c7e9013","resourceVersion":"137761"}},{"endpoint":"172.17.0.32:3000","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-4mot1","uid":"933d82a7-dc93-11e4-88b2-a0481c7e9013","resourceVersion":"253562"}}],"subsets":[{"addresses":[{"IP":"172.17.0.26","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-k5zf8","uid":"89cc6848-d7fd-11e4-88b2-a0481c7e9013","resourceVersion":"6422"}},{"IP":"172.17.0.30","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-7muj6","uid":"26e07ca5-da6d-11e4-88b2-a0481c7e9013","resourceVersion":"137761"}},{"IP":"172.17.0.32","targetRef":{"kind":"Pod","namespace":"default","name":"guestbook-controller-4mot1","uid":"933d82a7-dc93-11e4-88b2-a0481c7e9013","resourceVersion":"253562"}}],"ports":[{"port":3000,"protocol":"TCP"}]}]}}
```

This result includes updates for services "kubernetes", "kubernetes-ro", "redis-slave", "redis-master", etc.  Even though the requesting URL indicated only "guestbook" updates were to be watched.

This seems to only occur when no resourceVersion is specified and the most recent resourceVersion has fallen off the etcd watch window.

# Expected Behavior
The expected behavior when resourceVersion is not specified I would think should be either:
  1.  an initial update with the latest version of the resource, followed by filtered updates.
  2.  nothing, followed by filtered updates.

Though I'm not sure which of these is actually the spec.  I would prefer (1) since it allows you to do a resource watch without having to first do a separate GET.  But (2) coupled with a preceding GET would yield the same results.

**Comment by bgrant0607** [comment]
IRC: https://botbot.me/freenode/google-containers/2015-04-06/?msg=35967134&page=6

cc @lavalamp @brendandburns

**Comment by bgrant0607** [comment]
The pods mentioned in the stream are references to the pods corresponding to endpoints. See #4482 for discussion behind the feature.

We probably need to distinguish the resourceVersion at which an object was last modified, which is also used as an update precondition, vs. the current (global) resourceVersion needed for watch.

**Comment by bgrant0607** [comment]
@smarterclayton Does the change to use GET with watch as a query parameter enable the GET to both return the latest version and subscribe to changes, or is a separate non-watch GET still required?

**Comment by smarterclayton** [comment]
If you pass resourceVersion=0 you get list then watch (for single or multiple).

**Comment by smarterclayton** [comment]
That behavior has been there for a while - the only challenge is you don't know when the list is complete, and we felt that list then watch more clearly delineated that.

**Comment by lavalamp** [comment]
> We probably need to distinguish the resourceVersion at which an object was last modified, which is also used as an update precondition, vs. the current (global) resourceVersion needed for watch.

This.

**Comment by lavalamp** [comment]
Watching with resourceVersion=0 should return an error, IMO.

**Comment by invino4** [comment]
I think the semantics for {watch a single resource} vs. {watch all resources of the same type} should be separated.  

# Single Resource
When watching a single resource I would expect the watch to have *consistent* semantics.  That is, I would expect it to be possible to watch an item and be guaranteed either:
  1.  I saw every change to the item, or
  2.  I knew that a change was missed (and so could do a raw GET to re-sync).

To this end it would be ideal if watch for a single item had the following semantics:

## Basic Watch

```
curl http://localhost:8080/api/v1beta1/watch/endpoints/guestbook?namespace=default
```

Delivered the current state, followed by a list of all changes.  If any change is missed or can't be delivered then the watch is terminated.

## Reconnect Watch

```
curl http://localhost:8080/api/v1beta1/watch/endpoints/guestbook?namespace=default&resourceVersion=253567
```

Delivers either a guarantee of all changes that occurred on that resource from 253567 (exclusive) to present.  If the server cannot prove that there were no missing changes then it should return an error and terminate the watch.

This would imply that:

```
curl http://localhost:8080/api/v1beta1/watch/endpoints/guestbook?namespace=default&resourceVersion=0
```

Would always return an error unless the server can prove that it can deliver all changes to the resource (e.g. when the resource hasn't been created yet.)

# Multi-Resource Watch
I'm surprised that the interface wants to provide multi-resource watch.  It would seem that the semantics of this watch would either be potentially expensive to deliver or "best effort" such that it is basically useless for anything other than casual human consumption.

Imagine a future where the master is sharded such that even different resources of the same type are on different shards.  To provide meaningful semantics to a multi-resource watch would require that resource versions be meaningful in a global sense and in a way that spans all shards.  Wouldn't this be undesirable?  It would require that any write at any shard would need a round-trip message to all other shards to ensure that the logical clocks remain in sync.  A multi-watch would similarly have to be fanned out to all shards in the backend.  Both of these would defeat the benefits of sharding the data in the first place.

I could understand providing a multi-resource watch for the creation and deletion of resources of a particular type (handled potentially by a secondary index in the sharded scheme) without providing a full watch of the contents.  You could then use this watch to establish resource-specific watches for the resources you needed to track the contents of.  Maintaining a secondary index wouldn't have to be transactional with changes that produced or destroyed items.

**Comment by smarterclayton** [comment]
resourceVersion=0 is the same thing as saying watch resourceVersion=ObjectModifiedResourceVersion-1.  Perhaps it should be more explicit, but it is generally useful.  Why is blind watch bad?

----- Original Message -----

**Comment by smarterclayton** [comment]
Also, I don't see it mentioned above but `watch/<resourceType>/*` - everything in * is ignored.  Confusing, but already fixed by the deprecation of `watch`.

**Comment by bgrant0607** [comment]
cc @hchoudh

**Comment by smarterclayton** [comment]
We don't offer multi-resource watch today, but in the future you're likely to correlate very strongly with a shard based on your namespace, rather than a shard on your type.  I don't think sharding by type will be common except on a few types of resources (events, namespaces, possibly endpoints).

The access pattern for most clients is going to be "in namespace Foo, get type X, get type Y, get type Z, get type A".  Most UIs that want to watch anything will watch on the namespace.  So the access pattern is less likely to reward sharding by type.

**Comment by invino4** [comment]
> We don't offer multi-resource watch today.

The query:

```
http://localhost:8080/api/v1beta1/watch/endpoints
```

Is a multi-resource watch.  It watches for all resources of type Endpoints.

> future you're likely to correlate very strongly with a shard based on your namespace
But do you want the interface to **guarantee that all objects in the same namespace are updated with consistency**?  

I agree that shard boundaries are almost certainly NOT type-boundaries.  For instance there is clearly a consistency relationship between a Service and its Endpoint object.  I would hope that it would be impossible for me to observe a change to a Service object and then to subsequent query for its Endpoint object and get a stale read.  But this is a query across type.  Suggesting that the appropriate shard boundaries are certainly not type-boundaries.

**Comment by smarterclayton** [comment]
Sorry, I was referring to "multi-resource" as "watching over multiple resource *types*".  It is possible to emulate a watch across a small number of stable shards by creating a resource version tuple (a vector clock, essentially) - for medium sized clusters I see that as a reasonable option.  It's more likely though that infrastructure components that want to "watch the world" are either themselves shardable (such that it's acceptable for world to be defined as "my shard") or an intermediate component supports them.  That intermediate component might be something like a partially ordered event stream that provides a firehose to the cluster as an aggregator of the cluster changes.

Today, we require that API consumers assume resource version is internally consistent across a single type in a single namespace.  Controllers are allowed to watch the world until such a time as we implement a sharding solution.

There is no consistency relationship across resource types - services and endpoints are independent objects with independent lifecycles.  It is by design that Endpoints can be updated out of band with Services, to allow downstream components to offload them.  We relax referential rules across resource types - there is no guarantee a referenced object will be created before or after you, or that it should exist only if you exist.

**Comment by invino4** [comment]
I completely agree there are possible implementations that can tolerate sharded and/or inconsistent queries.  And deciding that certain kinds of JOINs, particularly across resource types and/or namespaces, are always going to be inconsistent seems like a fine place to land.  I was merely surprised that the API was designed to promote the appearance of consistency for these queries today.  I would have expected that the API would be more conservative to reserve these degrees of freedom around consistency for future implementations without requiring a future breaking change.  Offering less functionality in V1 or functionality with weaker semantics would allow weakly consistent future implementations to emerge later without breaking existing applications.

As a general rule I try to avoid APIs that give the appearance of consistency if they are not intended to actually deliver consistency.  I think most developers will always assume consistency unless there is a really good reason to believe otherwise.  This seems like human nature since people expect systems to exhibit the same causal behavior as the real world.

Take for instance Endpoints.  My impression of this after seeing the API was that Endpoints was simply a VIEW (in the SQL since) that provides a convenient way to express the JOIN between Service and Pods.  From that perspective I would certainly have expected there to be a causal relationship between a change in the Service and its corresponding Endpoints object.  

I could easily imagine myself making that mistake when writing a application.  Perhaps my program changes the Service object and then wants to send a message to all of the Endpoints associated with the new Service definition.  After changing the Service, my program queries Endpoints.  Did it get the new set of Endpoints?  There are several possible outcomes:

  1.  The Endpoints is the same as it was before I changed the Service Definition:
    1.  The new Service definition just happens to result in the same set of Endpoints as before.
    2.  I have not waited long enough for Endpoints to be updated.  How long should I wait?  What indicates that I've waited long enough?
  2.  The Endpoints is different:
    1.  Is this the correct one?  The one that corresponds to my Service definition change?  
    2.  Or is this the update that occurred due to a different Service definition change that happened just before mine? If so, I should wait longer for mine.
    3.  Or is this the update that occurred due to a different Service definition change that happened just after mine?  If so, I should either use this one or attempt my Service definition change again.

In an inconsistent and asynchronously updated world we need to provide a way to answer these questions.  (e.g. perhaps Endpoints should include the the resourceVersion of the Service definition used to compute it.)  All applications need to properly deal with this extra information.  (e.g. need to ignore all updated Endpoints until one arrives that has a resourceVersion >= their change to the Service definition.)  I'd imagine it's possible to implement this and for customers to use it correctly, but I think it offers a lot of opportunity for bugs.  Users might simply wait for the next update and then take the wrong action leading to a misbehaving system.  

If we could decide which relationships should exhibit consistency and provide it where the developer would expect it then the API is simply and easier to use and applications are more likely to be correct.  If we don't offer APIs which might lead the developer to make incorrect assumptions then applications are less likely to have those bugs.

**Comment by lavalamp** [comment]
I dug into this... it's more broken than I thought it was, so I've made a fix in #7263.

Once my fix goes in,
1. Watching a single object acts as if you didn't pass a name: this will be fixed directly (at least for the resource types I checked).
2. To get an appropriate resource version, you will have to LIST with a fieldselector ("?fields=name%3Dfoo" or some such-- don't recall exact syntax) instead of GETting. This way we don't have to modify the API to pass a special "watch at" resource version.
